File: frontend/src/api/reviews.ts
/**
 * Reviews API client
 */

import api from './axios'
import { getToken } from './auth'

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1'

export interface Review {
  id: number
  user_id: number
  game_id: number
  username: string
  rating: number
  comment: string
}

export interface ReviewCreate {
  rating: number
  comment: string
}

function getHeaders() {
  const token = getToken()
  return {
    'Content-Type': 'application/json',
    ...(token && { Authorization: `Bearer ${token}` }),
  }
}

export async function getGameReviews(gameId: number): Promise<Review[]> {
  const response = await api
    .get(`${API_URL}/games/${gameId}/reviews`, {
      headers: getHeaders(),
    })
    .catch(function (error) {
      throw new Error('Failed to fetch reviews', error)
    })

  return response.data
}

export async function createOrUpdateGameReview(
  gameId: number,
  review: ReviewCreate,
): Promise<Review> {
  const response = await api
    .post(`${API_URL}/games/${gameId}/reviews`, review, {
      headers: getHeaders(),
    })
    .catch(function (error) {
      throw new Error('Failed to submit review', error)
    })

  return response.data
}

File: backend/app/api/v1/reviews.py
"""API for game reviews."""

from fastapi import APIRouter, Depends, Header, HTTPException
from sqlalchemy.orm import Session

from app.core.security import decode_access_token
from app.db.session import get_db
from app.models.user import UserReview
from app.schemas.review_schema import ReviewCreate, ReviewResponse
from app.services.review_service import ReviewService

router = APIRouter(prefix="/games", tags=["reviews"])


def get_current_user_id(authorization: str = Header(None)) -> int:
    """Extract the user ID from the bearer token."""
    if not authorization:
        raise HTTPException(status_code=401, detail="Missing authorization header")

    token = authorization.replace("Bearer ", "")
    payload = decode_access_token(token)
    if not isinstance(payload, dict):
        raise HTTPException(status_code=401, detail="Invalid token")

    sub = payload.get("sub")
    if sub is None:
        raise HTTPException(status_code=401, detail="Invalid token payload")

    return int(sub)


def _serialize_review(review: UserReview) -> ReviewResponse:
    """Convert a review ORM object into the API response schema."""
    username = review.user.username if review.user else "Unknown"
    return ReviewResponse(
        id=review.id,
        user_id=review.user_id,
        game_id=review.game_id,
        username=username,
        rating=review.rating,
        comment=review.comment,
    )


@router.get("/{game_id}/reviews", response_model=list[ReviewResponse])
def get_game_reviews(game_id: int, db: Session = Depends(get_db)):
    """Get all reviews for a game."""
    reviews = ReviewService.get_game_reviews(db, game_id)
    return [_serialize_review(review) for review in reviews]


@router.post("/{game_id}/reviews", response_model=ReviewResponse)
def create_or_update_review(
    game_id: int,
    review: ReviewCreate,
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """Create or update the current user's review for a game."""
    saved_review = ReviewService.save_review(db, user_id, game_id, review)
    return _serialize_review(saved_review)
